In [13]:
import tweepy
import json
import pandas as pd

Authenticate with Twitter API

In [180]:
# twitter app credentials
CONSUMER_KEY = 'SZsWG8br2fSPUF1Xwd53iakd7'
CONSUMER_SECRET = '8CBYWeEMjdTML9xBRRDPhV6PwEFhKtHf92ye2PhMstYnC5Nhc5'
ACCESS_TOKEN = '404374326-ZaQe9uhMID24e7cPQDleLXB3FPFFGFytFZLCm7HR'
ACCESS_TOKEN_SECRET = 'Bb0LhjPmjyPYF3gAnDqTUcyDBUQCs13Ecb4yjZ4ni4R6G'
BEARER_TOKEN = 'AAAAAAAAAAAAAAAAAAAAABKSOQEAAAAAz%2BwUu%2F9HBgvVJHgw7QGiOKTfVtQ%3Dza18UIIgrQzYhzuO3FdFhANPjrpN5ozLZfDFqdI5O8KXwwcHDI'


auth = tweepy.OAuthHandler(CONSUMER_KEY, CONSUMER_SECRET)
auth.set_access_token(ACCESS_TOKEN, ACCESS_TOKEN_SECRET)

api = tweepy.API(auth, wait_on_rate_limit=True, wait_on_rate_limit_notify=True)

In [166]:
geo='48.136353,11.575004,25km'
COLS = ['id','created_at','lang','original text','user_name', 'place', 'place type', 'bbx', 'coordinates']

In [178]:
def run_test_query(keyword):
    df = pd.DataFrame(columns=COLS)

    for tweet in tweepy.Cursor(api.search,
                               q=keyword, 
                               include_rts=False,
                               geocode=geo,
                               lang="en").items(200):
        #creating string array
        new_entry = []

        #storing all JSON data from twitter API
        tweet = tweet._json    

        #Append the JSON parsed data to the string list:

        new_entry += [tweet['id'], tweet['created_at'], tweet['lang'], tweet['text'], 
                      tweet['user']['name']]

        #check if place name is available, in case not the entry is named 'no place'
        try:
            place = tweet['place']['name']
        except TypeError:
            place = 'no place'
        new_entry.append(place)

        try:
            place_type = tweet['place']['place_type']
        except TypeError:
            place_type = 'na'
        new_entry.append(place_type)

        try:
            bbx = tweet['place']['bounding_box']['coordinates']
        except TypeError:
            bbx = 'na'
        new_entry.append(bbx)

        #check if coordinates is available, in case not the entry is named 'no coordinates'
        try:
            coord = tweet['coordinates']['coordinates']
        except TypeError:
            coord = 'no coordinates'
        new_entry.append(coord)

        # wrap up all the data into a data frame
        single_tweet_df = pd.DataFrame([new_entry], columns=COLS)
        df = df.append(single_tweet_df, ignore_index=True)

        #get rid of tweets without a place
        df_cleaned = df[df.place != 'no place']
    
    df_cleaned.to_csv('tweets_'+geo+'.csv', columns=COLS,index=False)

In [179]:
keyword='*'
run_test_query(keyword)

Rate limit reached. Sleeping for: 718


In [154]:
for tweet in tweepy.Cursor(api.search,q="Womens Day OR internationalwomensday OR internationalwomensday2021 OR internationalwomensday21 OR women's day OR international women's day OR IWD OR HappyInternationalWomensDay OR Happy Women's Day OR HappyWomensDay OR happywomensday OR happyinternationalwomensday",
                           count=10000,
                           include_rts=False,
                           country_code=True,
                           coordinates=True,
                           lang="en",
                           since="2021-03-06",
                           until="2021-03-10"
                           ).items(10):
    print (tweet.created_at, tweet.text)

In [147]:
public_tweets = api.home_timeline()
for tweet in public_tweets:
    print(tweet.text)

The Last of Us Part II is $29.99 on US PSN https://t.co/EV48EuPKbg https://t.co/cYNPc1lZ0P
Xbox Wireless Controller – Cyberpunk 2077 Limited Edition is $62.82 on Amazon Warehouse (discount at checkout, Like… https://t.co/vKjVBRQmpL
We absolutely need to raise the minimum wage. But workers need additional protections, too.
https://t.co/26dcUpCk8w
A non-partisan ethics watchdog has asked the Senate Ethics Committee to launch an investigation into a traffic stop… https://t.co/zlmXzapCO3
Ask Iwata also $17.24 at Walmart https://t.co/F8eB3TzfKd https://t.co/4IQS3hcGNT
Age of Empires franchise sale on Steam https://t.co/cTvpc9uYta

also on Game Pass https://t.co/1cJXy2mRYk
Watch out for  r̶o̶a̶d̶ ̶b̶l̶o̶c̶k̶s̶ load outs, life may throw at you... #PS5Share #CallofDutyWarzone https://t.co/jZGIaGI2ZG
Balan Wonderworld (Steam) is $47.99 on GMG XP Offer or code SPR21 https://t.co/yGuRC210Y2
RT @amprog: Biden has prepared an executive order requiring federal contractors to pay a $15 minimum wage a

### parsing the sample tweet data Laura provided

In [85]:
us_state_abbrev = {
    'Alabama': 'AL',
    'Alaska': 'AK',
    'American Samoa': 'AS',
    'Arizona': 'AZ',
    'Arkansas': 'AR',
    'California': 'CA',
    'Colorado': 'CO',
    'Connecticut': 'CT',
    'Delaware': 'DE',
    'District of Columbia': 'DC',
    'Florida': 'FL',
    'Georgia': 'GA',
    'Guam': 'GU',
    'Hawaii': 'HI',
    'Idaho': 'ID',
    'Illinois': 'IL',
    'Indiana': 'IN',
    'Iowa': 'IA',
    'Kansas': 'KS',
    'Kentucky': 'KY',
    'Louisiana': 'LA',
    'Maine': 'ME',
    'Maryland': 'MD',
    'Massachusetts': 'MA',
    'Michigan': 'MI',
    'Minnesota': 'MN',
    'Mississippi': 'MS',
    'Missouri': 'MO',
    'Montana': 'MT',
    'Nebraska': 'NE',
    'Nevada': 'NV',
    'New Hampshire': 'NH',
    'New Jersey': 'NJ',
    'New Mexico': 'NM',
    'New York': 'NY',
    'North Carolina': 'NC',
    'North Dakota': 'ND',
    'Northern Mariana Islands':'MP',
    'Ohio': 'OH',
    'Oklahoma': 'OK',
    'Oregon': 'OR',
    'Pennsylvania': 'PA',
    'Puerto Rico': 'PR',
    'Rhode Island': 'RI',
    'South Carolina': 'SC',
    'South Dakota': 'SD',
    'Tennessee': 'TN',
    'Texas': 'TX',
    'Utah': 'UT',
    'Vermont': 'VT',
    'Virgin Islands': 'VI',
    'Virginia': 'VA',
    'Washington': 'WA',
    'West Virginia': 'WV',
    'Wisconsin': 'WI',
    'Wyoming': 'WY'
}

In [9]:
sample_feed_path = '../../data/us_corpus/Social Feed Manager/geo-tweets1.json' 

sample_data = []
with open(sample_feed_path, 'r') as file: 
    for line in file:
        sample_data.append(json.loads(line))

In [82]:
def get_country_value_counts(file):
    ''' get the countries where the geotagged tweets are from'''
    df = pd.DataFrame([tweet['place']['country_code'] for tweet in file if tweet['place']])
    return df

In [83]:
def filter_for_us_tweets_only(tweets):
    '''get us tweets only'''
    new_tweets = []
    for tweet in tweets: 
        if tweet['place']:
            if tweet['place']['country_code'] == 'US':
                new_tweets.append(tweet)
    return new_tweets

In [121]:
def get_us_states(tweets):
    '''get US states'''
    return [tweet['place']['full_name'].split(', ')[1:] for tweet in tweets]

In [133]:
def change_place_name_to_abbrev(tweets):
    for tweet in tweets: 
        possible_full_state_name = tweet['place']['full_name'].split(', ')[0]
        if possible_full_state_name in us_state_abbrev.keys():
            tweet['place']['full_name'] = tweet['place']['full_name'].replace('USA', us_state_abbrev[possible_full_state_name])
    return tweets

In [134]:
testing = get_country_value_counts(sample_data)
us_tweets = filter_for_us_tweets_only(sample_data)
us_states = get_us_states(change_place_name_to_abbrev(us_tweets))

In [137]:
pd.DataFrame(us_states).value_counts()[:10]

CA    1539
NY    1287
FL     757
TX     719
NJ     406
GA     394
NC     334
PA     328
OH     322
IL     313
dtype: int64

In [32]:
testing.value_counts()

US    11069
GB     1840
CA      752
IN      684
NG      334
      ...  
NI        1
CI        1
PN        1
PY        1
KG        1
Length: 152, dtype: int64

In [33]:
11069/17983

0.6155257743424345

In [140]:
757/11069

0.0683891950492366

11069 / 17983 (61%) of the tweets are from US

Of these 11069 tweets from the US, 1539 are from CA (13%), 1287 (11%) are from NY, 757 (6.8%) are from FL. 

In [59]:
sample_data[5]['place']['full_name'].split(', ')[1]

'PA'

In [141]:
17983/94000000

0.00019130851063829787

In [144]:
758000000 * 0.00019130851063829787 * 0.61* 0.10

8845.722914893617